[Source compatibility suite] Result failing to build with main branch

### Description

The [Result](https://github.com/antitypical/Result/tree/master) project  is failing to build with the main branch. The error seen in the log is
```
Result/Result/Result.swift:87:14: error: ambiguous use of 'get()'
                return try get()
                           ^
Result/Result/Result.swift:95:14: note: found this candidate 
        public func get() throws -> Success {
                    ^
Swift.Result:8:28: note: found this candidate in module 'Swift'
    @inlinable public func get() throws(Failure) -> Success
                           ^
error: fatalError
```

This was seemingly introduced with https://github.com/apple/swift/commit/7d74b3ba5c0f66d630cee8652263979eeed11289

### Reproduction

```
swift build --disable-sandbox --package-path Result --verbose --configuration debug --build-tests -Xswiftc -swift-version -Xswiftc 5 -Xswiftc -enable-testing
```

### Expected behavior

Expect the project to build

### Environment

```swift
Apple Swift version 5.11-dev (LLVM 54cbee1d059ba29, Swift 467f684a430cff0)
Target: x86_64-apple-macosx12.0
```

### Additional information

https://ci.swift.org/view/Dashboard/job/swift-main-source-compat-suite-debug/653/

--- a/lib/AST/NameLookup.cpp
+++ b/lib/AST/NameLookup.cpp
@@ -733,6 +733,53 @@ static void recordShadowedDeclsAfterTypeMatch(
   }
 }
 
+/// Return an extended info for a function types that removes the use of
+/// the thrown error type, if present.
+///
+/// Returns \c None when no adjustment is needed.
+static std::optional<ASTExtInfo>
+extInfoRemovingThrownError(AnyFunctionType *fnType) {
+  if (!fnType->hasExtInfo())
+    return std::nullopt;
+
+  auto extInfo = fnType->getExtInfo();
+  if (!extInfo.isThrowing() || !extInfo.getThrownError())
+    return std::nullopt;
+
+  return extInfo.withThrows(true, Type());
+}
+
+/// Remove the thrown error type.
+static CanType removeThrownError(Type type) {
+  ASTContext &ctx = type->getASTContext();
+
+  return type.transformRec([](TypeBase *type) -> std::optional<Type> {
+    if (auto funcTy = dyn_cast<FunctionType>(type)) {
+      if (auto newExtInfo = extInfoRemovingThrownError(funcTy)) {
+        return FunctionType::get(
+                  funcTy->getParams(), funcTy->getResult(), *newExtInfo)
+          ->getCanonicalType();
+      }
+
+      return std::nullopt;
+    }
+
+    if (auto genericFuncTy = dyn_cast<GenericFunctionType>(type)) {
+      if (auto newExtInfo = extInfoRemovingThrownError(genericFuncTy)) {
+        return GenericFunctionType::get(
+                  genericFuncTy->getGenericSignature(),
+                  genericFuncTy->getParams(), genericFuncTy->getResult(),
+                  *newExtInfo)
+          ->getCanonicalType();
+      }
+
+      return std::nullopt;
+    }
+
+    return std::nullopt;
+  })->getCanonicalType();
+}
+
 /// Given a set of declarations whose names and generic signatures have matched,
 /// figure out which of these declarations have been shadowed by others.
 static void recordShadowedDeclsAfterSignatureMatch(
@@ -757,7 +804,7 @@ static void recordShadowedDeclsAfterSignatureMatch(
     if (auto asd = dyn_cast<AbstractStorageDecl>(decl))
       type = asd->getOverloadSignatureType();
     else
-      type = decl->getInterfaceType()->getCanonicalType();
+      type = removeThrownError(decl->getInterfaceType()->getCanonicalType());
 
     // Record this declaration based on its signature.
     auto &known = collisions[type];

--- a/test/decl/func/typed_throws.swift
+++ b/test/decl/func/typed_throws.swift
@@ -190,3 +190,20 @@ func testArrMap(arr: [String]) {
   _ = mapArray(arr, body: G_E<Int>.tuple)
   // expected-error@-1{{cannot convert value of type '((x: Int, y: Int)) -> G_E<Int>' to expected argument type '(String) -> G_E<Int>'}}
 }
+
+// Shadowing of typed-throws Result.get() addresses a source compatibility
+// issue with the introduction of typed throws.
+extension Result {
+  public func dematerialize() throws -> Success {
+    return try get()
+  }
+
+  public func get() throws -> Success {
+    switch self {
+    case let .success(value):
+      return value
+    case let .failure(error):
+      throw error
+    }
+  }
+}

